iquip/apps/builder.py

```diff
@@ -0,0 +1,206 @@
+        parent: Optional[QObject] = None
+    ):
+        """Extended.
+        
+        Args:
+            experimentPath: See the attributes section in ExperimentInfoThread.
+            callback: The callback method called after this thread is finished.
+        """
+        super().__init__(parent=parent)
+        self.experimentPath = experimentPath
+        self.fetched.connect(callback, type=Qt.QueuedConnection)
+
+    def run(self):
+        """Overridden.
+        
+        Fetches the experiment information from the proxy server.
+        
+        If the path is a directory, 500 Server error occurs.
+        If the path is a non-experiment file, the server returns an empty dictionary.
+
+        The experiment information is an instance of protocols.ExperimentInfo.
+        After finished, the fetched signal is emitted.
+        """
+        try:
+            response = requests.get("http://127.0.0.1:8000/experiment/info/",
+                                    params={"file": self.experimentPath},
+                                    timeout=10)
+            response.raise_for_status()
+            data = response.json()
+        except requests.exceptions.RequestException as err:
+            print(err)
+            return
+        if data:
+            experimentClsName = next(iter(data))
+            experimentInfo = data[experimentClsName]
+            self.fetched.emit(
+                self.experimentPath,
+                experimentClsName,
+                ExperimentInfo(**experimentInfo)
+            )
+
+
+class ExperimentSubmitThread(QThread):
+    """QThread for submitting the experiment with its build arguments.
+    
+    Signals:
+        submitted(rid): The experiment is submitted.
+    
+    Attributes:
+        experimentPath: The path of the experiment file.
+        experimentArgs: The arguments of the experiment.
+    """
+
+    submitted = pyqtSignal(int)
+
+    def __init__(
+        self,
+        experimentPath: str,
+        experimentArgs: Dict[str, Any],
+        callback: Callable[[int], None],
+        parent: Optional[QObject] = None
+    ):
+        """Extended.
+        
+        Args:
+            experimentPath, experimentArgs: See the attributes section in ExperimentSubmitThread.
+            callback: The callback method called after this thread is finished.
+        """
+        super().__init__(parent=parent)
+        self.experimentPath = experimentPath
+        self.experimentArgs = experimentArgs
+        self.submitted.connect(callback, type=Qt.QueuedConnection)
+
+    def run(self):
+        """Overridden.
+        
+        Submits the experiment to the proxy server.
+
+        Regardless of whether it is successful or not, the server returns the run identifier.
+        After finished, the submitted signal is emitted.
+        """
+        try:
+            params = {
+                "file": self.experimentPath,
+                "args": json.dumps(self.experimentArgs)
+            }
```

Comment:
I think this should be outside this `try:` block.
Maybe you should use another `try:` block and catch the exception from `json.dumps()`.

Reply:
It looks better!